From 118de212d38232ba7f3ddebcb000bc1cdcb28707 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20Dr=C3=B6ge?= <sebastian@centricular.com>
Date: Wed, 7 May 2014 19:33:43 +0200
Subject: [PATCH 3/3] Pop the thread default main context not from the callback
 but from where it was pushed

Based on a patch by Dmitry Shatrov <shatrov@gmail.com>, see
http://mail.gnome.org/archives/libsoup-list/2011-November/msg00000.html

https://bugzilla.gnome.org/show_bug.cgi?id=729737
---
 libsoup/soup-connection.c | 11 +++++------
 libsoup/soup-socket.c     |  7 +++----
 2 files changed, 8 insertions(+), 10 deletions(-)

diff --git a/libsoup/soup-connection.c b/libsoup/soup-connection.c
index fce589f..9d8afbc 100644
--- a/libsoup/soup-connection.c
+++ b/libsoup/soup-connection.c
@@ -448,9 +448,6 @@ socket_connect_finished (GTask *task, SoupSocket *sock, GError *error)
 	SoupConnection *conn = g_task_get_source_object (task);
 	SoupConnectionPrivate *priv = SOUP_CONNECTION_GET_PRIVATE (conn);
 
-	if (priv->async_context && !priv->use_thread_context)
-		g_main_context_pop_thread_default (priv->async_context);
-
 	g_signal_handlers_disconnect_by_func (sock, G_CALLBACK (re_emit_socket_event), conn);
 
 	if (!error) {
@@ -563,6 +560,9 @@ soup_connection_connect_async (SoupConnection      *conn,
 
 	soup_socket_connect_async_internal (priv->socket, cancellable,
 					    socket_connect_complete, task);
+	if (priv->async_context && !priv->use_thread_context)
+		g_main_context_pop_thread_default (priv->async_context);
+
 }
 
 gboolean
@@ -689,9 +689,6 @@ start_ssl_completed (GObject *object, GAsyncResult *result, gpointer user_data)
 	SoupConnectionPrivate *priv = SOUP_CONNECTION_GET_PRIVATE (conn);
 	GError *error = NULL;
 
-	if (priv->async_context && !priv->use_thread_context)
-		g_main_context_pop_thread_default (priv->async_context);
-
 	if (soup_socket_handshake_finish (priv->socket, result, &error)) {
 		soup_connection_event (conn, G_SOCKET_CLIENT_TLS_HANDSHAKED, NULL);
 		soup_connection_event (conn, G_SOCKET_CLIENT_COMPLETE, NULL);
@@ -721,6 +718,8 @@ soup_connection_start_ssl_async (SoupConnection      *conn,
 
 	soup_socket_handshake_async (priv->socket, priv->remote_uri->host,
 				     cancellable, start_ssl_completed, task);
+	if (priv->async_context && !priv->use_thread_context)
+		g_main_context_pop_thread_default (priv->async_context);
 }
 
 gboolean
diff --git a/libsoup/soup-socket.c b/libsoup/soup-socket.c
index b9f1dfc..dced4dd 100644
--- a/libsoup/soup-socket.c
+++ b/libsoup/soup-socket.c
@@ -799,14 +799,10 @@ legacy_connect_async_cb (GObject       *object,
 			 gpointer       user_data)
 {
 	SoupSocket *sock = SOUP_SOCKET (object);
-	SoupSocketPrivate *priv = SOUP_SOCKET_GET_PRIVATE (sock);
 	SoupSocketAsyncConnectData *sacd = user_data;
 	GError *error = NULL;
 	guint status;
 
-	if (priv->async_context && !priv->use_thread_context)
-		g_main_context_pop_thread_default (priv->async_context);
-
 	if (soup_socket_connect_finish_internal (sock, result, &error))
 		status = SOUP_STATUS_OK;
 	else
@@ -856,6 +852,9 @@ soup_socket_connect_async (SoupSocket *sock, GCancellable *cancellable,
 	soup_socket_connect_async_internal (sock, cancellable,
 					    legacy_connect_async_cb,
 					    sacd);
+	if (priv->async_context && !priv->use_thread_context)
+		g_main_context_pop_thread_default (priv->async_context);
+
 }
 
 gboolean
-- 
1.9.0

